Desired Manifest should be persisted (versioned) in the var interpolation step

edwardstudy opened the pull request:
https://www.pivotaltracker.com/story/show/165020388

viovanov commented on pkg/kube/controllers/boshdeployment/manifest_reconciler.go:
```diff
@@ -0,0 +1,126 @@
+	ctxlog.Infof(ctx, "Reconciling BOSHDeployment %s for desired manifest creation", request.NamespacedName)
+
+	// Fetch the BOSHDeployment instance
+	instance := &bdv1.BOSHDeployment{}
```
Isn't this reconciling a `Secret`?

viovanov commented on pkg/kube/controllers/extendedjob/job_reconciler.go:
```diff
@@ -19,19 +19,25 @@ import (
 
 // NewJobReconciler returns a new Reconciler
 func NewJobReconciler(ctx context.Context, config *config.Config, mgr manager.Manager, podLogGetter PodLogGetter) (reconcile.Reconciler, error) {
+	manifestStore := store.NewStore(mgr.GetClient(), bdm.DesiredManifestKeyName)
```
This is not ok - the `ExtendedJob` should have no dependencies on "bosh manifest" stuff.

viovanov commented on pkg/kube/controllers/extendedjob/job_reconciler.go:
```diff
@@ -178,27 +185,50 @@ func (r *ReconcileJob) persistOutput(ctx context.Context, instance *batchv1.Job,
-		err = r.client.Get(ctx, types.NamespacedName{Name: secretName, Namespace: instance.GetNamespace()}, &corev1.Secret{})
-		if apierrors.IsNotFound(err) {
-			err = r.client.Create(ctx, secret)
+		if conf.ToBeVersioned {
```
We don't want the `ExtendedJob` to do versioning.
It's fine if the variable interpolation `ExtendedJob` saves output to a normal secret, and the **manifest reconciler** picks that up, and it then versions it.

We don't want any of the `Extended*` resources to be dependent on BOSH concepts.

edwardstudy replied:
So what **manifest reconciler** does is that watch new desired manifest from var-interpolation job and store as versioned manifest. So it just archives each manifest and **does not** apply versioned manifest to data-gathering job input. Is it right?

edwardstudy commented on pkg/bosh/manifest/kube_converter.go:
```diff
@@ -178,18 +179,17 @@ func (m *Manifest) variableInterpolationJob(namespace string) (*ejv1.ExtendedJob
 			Name:      eJobName,
 			Namespace: namespace,
 			Labels: map[string]string{
-				bdv1.LabelKind:       "variable-interpolation",
```
I just cleaned up labels/annotations under `apis`. It's useless so I removed it.

edwardstudy commented on pkg/kube/controllers/boshdeployment/deployment_reconciler.go:
```diff
@@ -270,7 +269,7 @@ func (r *ReconcileBOSHDeployment) applyOps(ctx context.Context, instance *bdv1.B
 	// retrieve manifest
 	ctxlog.Debug(ctx, "Resolving manifest")
-	manifest, err := r.resolver.ResolveManifest(instance.Spec, instance.GetNamespace())
+	manifest, err := r.resolver.ResolveManifest(instance.Spec, instance.GetNamespace(), instance.GetName())
```
Got it

edwardstudy commented:
Hi, @viovanov 
In the last commit, I changed lots of things.

1. Removed global function in `bosh/manifest` package
2. Decoupled BoshDeployment reconciler with versioned secret changes and moved it to each extendedJobs/extendedStatefulSet reconciler: 
    * In extendedStatefulSet reconciler, it will watch versioned secrets and update extendedStatefulSets reference(envs, volumes)
    * In extendedJobs reconciler, it will do the same thing but split two reconcilers: owner reconciler and errand reconciler


Please help to review. Thanks!
Edward

viovanov commented on pkg/kube/controllers/extendedsecret/versioning.go:
```diff
@@ -0,0 +1,260 @@
+package extendedsecret
```
this should live in `pkg/kube/versioned_secret_store/`

viovanov commented on pkg/kube/controllers/extendedstatefulset/reconciler.go:
```diff
@@ -513,6 +525,58 @@ func (r *ReconcileExtendedStatefulSet) updateStatefulSetsConfigSHA1(ctx context.
+		secret := &corev1.Secret{}
+		err = r.client.Get(ctx, secretKey, secret)
+		if err != nil && apierrors.IsNotFound(err) {
+			// Replace secret ref with a new versioned secret
```
If the secret is not found, how do you know it's a versioned secret?

viovanov commented on pkg/kube/controllers/extendedstatefulset/reconciler.go:
```diff
@@ -513,6 +525,58 @@ func (r *ReconcileExtendedStatefulSet) updateStatefulSetsConfigSHA1(ctx context.
+		if secretLabels == nil {
+			continue
+		}
+		referencedSecretName := secretLabels[ejv1.LabelReferencedSecretName]
```
What is `ejv1.LabelReferencedSecretName`?

viovanov commented on pkg/kube/apis/extendedjob/v1alpha1/types.go:
```diff
@@ -1,16 +1,26 @@
+	// LabelReferencedJobName is the name key for dependent job
+	LabelReferencedJobName = fmt.Sprintf("%s/referenced-job-name", apis.GroupName)
+	// LabelReferencedSecretName is the name key for dependant's referenced secret
+	LabelReferencedSecretName = fmt.Sprintf("%s/referenced-secret-name", apis.GroupName)
```
We shouldn't need this anymore

edwardstudy replied:
I thought we use it to fetch the latest version when update versioned secret in reconciler.

viovanov commented on pkg/kube/controllers/extendedjob/errand_reconciler.go:
```diff
@@ -166,3 +181,62 @@ func (r *ErrandReconciler) createJob(ctx context.Context, extJob ejv1.ExtendedJo
 }
+
+// updateSecretReference look and see if any referenced secret is versioned, and if it is, but it's not latest, you have to change the reference
+func (r *ErrandReconciler) updateSecretReference(ctx context.Context, extJob *ejv1.ExtendedJob) error {
```
This pattern is defined in multiple places.
Can we move it to the versioned secret store?

edwardstudy replied:
Yes. It should be.

viovanov commented on pkg/kube/controllers/extendedjob/errand_reconciler.go:
```diff
@@ -166,3 +181,62 @@ func (r *ErrandReconciler) createJob(ctx context.Context, extJob ejv1.ExtendedJo
+		}
+	}
+
+	if extJobCopy != extJob {
```
I think you need `DeepEqual` here - otherwise this is always false
